cli_tools/gce_windows_upgrade/upgrader/utils.go

```diff
@@ -0,0 +1,146 @@
+//  Copyright 2020 Google Inc. All Rights Reserved.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package upgrader
+
+import (
+	"bytes"
+	"strings"
+	"text/template"
+
+	daisyutils "github.com/GoogleCloudPlatform/compute-image-tools/cli_tools/common/utils/daisy"
+	"github.com/GoogleCloudPlatform/compute-image-tools/cli_tools/common/utils/param"
+	"github.com/GoogleCloudPlatform/compute-image-tools/cli_tools/daisycommon"
+	"github.com/GoogleCloudPlatform/compute-image-tools/daisy"
+)
+
+const (
+	upgradeIntroductionTemplate = "The following resources will be created/accessed during the upgrade. " +
+		"Please keep track of their names if manual cleanup and/or rollback are necessary.\n" +
+		"All resources are in project '{{.project}}', zone '{{.zone}}'.\n" +
+		"1. Instance: {{.instanceName}}\n" +
+		"2. Disk for install media: {{.installMediaDiskName}}\n" +
+		"3. Snapshot for original OS disk: {{.osDiskSnapshotName}}\n" +
+		"4. Original OS disk: {{.osDiskName}}\n" +
+		"   - Device name of the attachment: {{.osDiskDeviceName}}\n" +
+		"   - AutoDelete of the attachment: {{.osDiskAutoDelete}}\n" +
+		"5. New OS disk: {{.newOSDiskName}}\n" +
+		"6. Machine image: {{.machineImageName}}\n" +
+		"7. Original startup script url: {{.originalStartupScriptURL}}\n" +
+		"\n" +
+		"When upgrading succeeded but cleanup failed, please manually cleanup by following steps:\n" +
+		"1. Delete 'windows-startup-script-url' from the instance's metadata if there isn't an original value. " +
+		"If there is an original value, restore it. The original value is backed up as metadata 'windows-startup-script-url-backup'.\n" +
+		"2. Detach the install media disk from the instance and delete it.\n" +
+		"\n" +
+		"When upgrading failed but you didn't enable auto-rollback, or auto-rollback failed, or " +
+		"upgrading succeeded but you still need to rollback for any reason, " +
+		"please manually rollback by following steps:\n" +
+		"1. Detach the new OS disk from the instance and delete it.\n" +
+		"2. Attach the old OS disk as boot disk.\n" +
+		"3. Detach the install media disk from the instance and delete it.\n" +
+		"4. Delete 'windows-startup-script-url' from the instance's metadata if there isn't an original value. " +
+		"If there is an original value, restore it. The original value is backed up as metadata 'windows-startup-script-url-backup'.\n" +
```

Comment:
new text: "If there is an original value for the script, restore the value. The original value is backed up as metadata 'windows-startup-script-url-backup'."